# Changelog

All notable changes to this project will be documented in this file.

## [1.1.3] - 2024-11-26

### Added

- Added `--force-samples` when subsetting based on sample IDs.

### Fixed

- Fixed the python file to ensure precise variant filtration based on chromosome, position, reference, and alternate alleles.
- Error: `bcftools: error while loading shared libraries: libcrypto.so.1.0.0: cannot open shared object file: No such file or directory`

## [1.1.2] - 2024-04-03

### Added

- Added a new boolean input variable "match_by_pos_only" to allow matching variants based on position only.
- Added conditional execution of commands in the loop based on the "match_by_pos_only" flag. If "match_by_pos_only" is true, a tab-delimited file is created with the chromosome, position, reference allele, and alternate allele from the subset VCF file. The exact chromosome, position, reference allele, and alternate allele from the newly created file are then matched with the subset variants list, and a final VCF file with the matched alleles is created.
- Updated the "VCF_FILES" variable to include the correct file name based on the "match_by_pos_only" condition.

## [1.1.1] - 2024-04-02

### Added

- The GT encoding now follows a custom order for the genotypes. Both 0/1 and 1/0 are represented in the same order: Ref/Alt. 0/0 is represented as Ref/Ref, and 1/1 is represented as Alt/Alt.
- Users now have the option to switch between the desired genotype encoding by using the "use_GT_from_PED" input flag. By setting "use_GT_from_PED" to true, the genotype encoding will be sourced from the PED file generated by Plink2 software.

## [1.0.1] - 2024-03-28

### Added

- Added support for genotype encoding using PLINK2 software
- The python script now has an optional --ped argument. If supplied, genotypes will be sourced from this file. Otherwise, genotypes will be extracted from the VCF file.

## [1.0.0] - 2024-03-27

### Added

- Initial release
